aws/retry/middleware.go

```diff
@@ -0,0 +1,177 @@
+package retry
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/awslabs/smithy-go/middleware"
+	"github.com/awslabs/smithy-go/transport/http"
+)
+
+type retryMetadata struct {
+	AttemptNum       int
+	AttemptTime      time.Time
+	MaxAttempts      int
+	AttemptClockSkew time.Duration
+}
+
+type retryMetadataKey struct{}
+
+type RetryMiddleware struct {
+	Retryer Retryer
+}
+
+func (r RetryMiddleware) ID() string {
+	return "Retry Middleware"
+}
+
+func (r RetryMiddleware) Name() string {
+	return r.ID()
+}
+
+func (r RetryMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (out middleware.FinalizeOutput, err error) {
+	const invocationIDHeader = "amz-sdk-invocation-id"
+
+	req := in.Request.(*http.Request)
+
+	// TODO: Don't see a compelling reason to make this a separate middleware that makes this available via context
+	invocationID, err := sdk.UUIDVersion4()
+	if err != nil {
+		return out, err
+	}
+
+	req.Header.Set(invocationIDHeader, invocationID)
+
+	var attemptNum, retryCount int
+	var attemptClockSkew time.Duration
+
+	maxAttempts := r.Retryer.MaxAttempts()
+
+	relRetryToken := r.Retryer.GetInitialToken()
+	for {
+		attemptNum++
+
+		ctx = context.WithValue(ctx, retryMetadataKey{}, retryMetadata{
+			AttemptNum:       attemptNum,
+			AttemptTime:      sdk.NowTime(),
+			MaxAttempts:      maxAttempts,
+			AttemptClockSkew: attemptClockSkew,
```

Comment:
Clock skew probably should be a separate middleware that determines this value?

Reply:
Agreed on calculating this as part of de-serialization and having someone to access it in the response.